src/include/libpmemobj.h

```diff
@@ -84,6 +99,553 @@ void pmemobj_set_funcs(
+int pmemobj_rwlock_trywrlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_unlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+
+int pmemobj_cond_zero(PMEMcond *condp);
+int pmemobj_cond_broadcast(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_signal(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_timedwait(PMEMobjpool *pop, PMEMcond *restrict condp,
+	PMEMmutex *restrict mutexp, const struct timespec *restrict abstime);
+int pmemobj_cond_wait(PMEMobjpool *pop, PMEMcond *condp,
+	PMEMmutex *restrict mutexp);
+
+/* find a heap offset of pool identified by uuid_lo */
+uint64_t pmemobj_heap_off_by_uuid_lo(uint64_t uuid_lo);
+
+/*
+ * Persistent memory object
+ */
+
+/*
+ * Object handle
+ */
+typedef struct pmemoid {
+	uint64_t pool_uuid_lo;
+	uint64_t off;
+} PMEMoid;
+
+#define	OID_TYPE(type)							\
+union {									\
+	type *__type;							\
+	PMEMoid oid;							\
+}
+
+#define	OID_ASSIGN(o, value)						\
+	((o).oid = value)
+
+#define	TYPED_OID_ASSIGN(lhs, rhs) (					\
+{(lhs).__type = (rhs).__type; (lhs).oid = (rhs).oid; })
+
+#define	OID_NULL		((PMEMoid) {0, 0})
+#define	MAX_OID_TYPE_NUM	1024
+
+#define	OID_IS_NULL(o)	((o).oid.off == 0)
+
+#define	OID_EQUALS(lhs, rhs)						\
+	((lhs).oid.off == (rhs).oid.off &&				\
+	(lhs).oid.pool_uuid_lo == (rhs).oid.pool_uuid_lo)
+
+#define	_DIRECT_OFF(o)							\
+	((o).oid.off + pmemobj_heap_off_by_uuid_lo((o).oid.pool_uuid_lo))
+
+#define	DIRECT_RW(o)	((typeof (*(o).__type)*)_DIRECT_OFF(o))
+#define	DIRECT_RO(o)	((const typeof (*(o).__type)*)_DIRECT_OFF(o))
+
+#define	D_RW	DIRECT_RW
+#define	D_RO	DIRECT_RO
+
+/*
+ * Non-transactional atomic allocations
+ *
+ * Those functions can be used outside transactions. The allocations are always
+ * aligned to the cache-line boundary.
+ */
+
+typedef void (*alloc_construct_func)(void *ptr, void *arg);
+
+/*
+ * Allocates a new object from the pool.
+ */
+PMEMoid pmemobj_alloc(PMEMobjpool *pop, size_t size, int type_num);
+
+/*
+ * Allocates a new zeroed object from the pool.
+ */
+PMEMoid pmemobj_zalloc(PMEMobjpool *pop, size_t size, int type_num);
+
+/*
+ * Allocates a new object from the pool and calls a constructor function before
+ * returning. It is guaranteed that allocated object is either properly
+ * initialized, or if it's interrupted before the constructor completes, the
+ * memory reserved for the object is automatically reclaimed.
+ */
+PMEMoid pmemobj_alloc_construct(PMEMobjpool *pop, size_t size, int type_num,
+	alloc_construct_func constructor, void *arg);
+
+/*
+ * Resizes an existing object.
+ */
+PMEMoid pmemobj_realloc(PMEMobjpool *pop, PMEMoid oid, size_t size,
+	int type_num);
+
+/*
+ * Resizes an existing object, if extended new space is zeroed.
+ */
+PMEMoid pmemobj_zrealloc(PMEMobjpool *pop, PMEMoid oid, size_t size,
+	int type_num);
+
+/*
+ * Allocates a new object with duplicate of the string s.
+ */
+PMEMoid pmemobj_strdup(PMEMobjpool *pop, const char *s, size_t n, int type_num);
```

Comment:
Frankly, I don't really see the use case for this function.
And by the way, this should be named *pmemobj_strndup* actually. The question is if we need also *pmemobj_strdup*, or only one of those.